Psychtoolbox/PsychHardware/EyelinkToolbox/EyelinkDemos/SR-ResearchDemos/MRI_BlockRecord/EyeLink_MRI_BlockRecord.m

```diff
@@ -0,0 +1,430 @@
+            % Prepare image on backbuffer
+            imgName = char(imgList{iBlock}{i});% Get image file name for current trial
+            imgInfo = imfinfo(imgName); % Get image file info
+            imgData = imread(imgName); % Read image from file
+            imgTexture = Screen('MakeTexture',window, imgData); % Convert image file to texture
+            Screen('DrawTexture', window, imgTexture); % Prepare image texture on backbuffer           
+            % Save complete backbuffer as trial*.bmp to be used as stimulus & DataViewer backdrop
+            stimName = ['blk' num2str(iBlock) 'trial' num2str(i) '.bmp']; % Prepare stimulus file name
+            stimArray = Screen('GetImage', window, [], 'drawBuffer'); % Copy backbuffer to be used as stimulus
+            imwrite(stimArray, stimName); % Save .bmp stimulus file in experment folder           
+            % Convert stimulus to texture so it is ready for drawing later on
+            stimTexture = Screen('MakeTexture', window, stimArray); % Convert to texture
+            
+            % STEP 7.2: START TRIAL AND DRAW CROSSHAIRS FOR DATAVIEWER
+            
+            if i ~= 1 % If not first trial in a block
+                % Write TRIALID message to EDF file: marks the start of a trial for DataViewer
+                % TRIALID message for first trial in the block was written before recording began
+                % See DataViewer manual section: Protocol for EyeLink Data to Viewer Integration > Defining the Start and End of a Trial
+                trialCount = trialCount + 1; % Add 1 to trial counter
+                Eyelink('Message', 'TRIALID %d', trialCount);
+            end
+            
+            % Write messages to EDF: prepare backdrop and draw central crosshairs for DataViewer
+            % See DataViewer manual section: Protocol for EyeLink Data to Viewer Integration > Simple Drawing
+            Eyelink('Message', '!V CLEAR %d %d %d', el.backgroundcolour(1), el.backgroundcolour(2), el.backgroundcolour(3));
+            Eyelink('Message', '!V DRAWLINE 0 0 0 %d %d %d %d', round(width/2-20), round(height/2), round(width/2+20), round(height/2));
+            Eyelink('Message', '!V DRAWLINE 0 0 0 %d %d %d %d', round(width/2), round(height/2-20), round(width/2), round(height/2+20));
+            
+            % STEP 7.3: WAIT FOR MRI TRIGGER AND PRESENT STIMULUS; CREATE DATAVIEWER STIMULUS BACKDROP AND INTEREST AREA
+            
+            % Present initial trial image
+            Screen('DrawTexture', window, stimTexture); % Prepare stimulus texture on backbuffer            
+            % Wait for MRI trigger (keyboard key "t" in this demo)
+            if i == 1 % If first trial in a block
+                KbReleaseWait; % Wait until user releases keys on keyboard:
+                keyTrigger = KbName('t');% Identify key code for keyboard key "t"
+                while 1
+                    [~, ~, keyCode] = KbCheck;

```

Comment:
So this is possibly a problem on anything but MS-Windows. Only MS-Windows can't distinguish between different connected "Keyboards". But Linux and macOS treat each connected keyboard - or some fMRI trigger gizmo that pretends to be a keyboard - as a different input device. Using `KbCheck` will only query the default keyboard and likely miss triggers from the scanner on Linux or macOS. Better to pass in a `triggerDevice` index a la ``KbCheck(triggerDevice)``, similar for ``KbReleaseWait(triggerDevice)`` etc. The proper device index can be found via the helper function ``triggerDevice =GetKeyboardIndices ('Name of the keyboard product');`` with a list of product names readily available via ``[~, names] = GetKeyboardIndices``

It would be good to explain this to users somehow. If you want a simple solution that at least would work, albeit not optimally, you could set `triggerDevice = -1;` which means to query all connected keyboards. This gives the same behaviour as on MS-Windows, with the downside that now 't' from the MRI scanner and 't' from the subjects keyboard or experimenters keyboard can no longer be distinguished.

Reply:
thank you for clarifying this behaviour. I will add this as a priority item to handle in the next round of improvements once this pull request is complete. We haven't received any complaints and I'd like to use additional time to understand the usage conventions for a triggerDevice, whether setting this up would best be done in the MRI demo or more generally EyelinkInit/UpdateDefaults wrapper, and that the changes work seamlessly alongside other intended keyboard keypress interactions with the experiment/participant